app/configuration.py

```diff
@@ -151,25 +163,23 @@ def path_to_yaml_data(path: str):
         return yaml.safe_load(file)
 
 
-def _interpolate_environment_vars(data: dict):
-    interpolated = {}
-    for key, value in data.items():
-        var_name = value
-        if value.startswith("$"):
-            var_name = value[1:]
-            if value == f"${key.upper()}":
-                value = os.getenv(var_name, "")
-                logging.debug(f"Found value {value!r} in env var {var_name}")
-            else:
+def _interpolate_environment_vars(config_value: str, config_key: str):
+    for var_name in ENV_VAR_KEYS:
+        match_str = "${" + var_name + "}"
+        if match_str in config_value:
+            env_value = os.getenv(var_name, "")
+            if not env_value:
                 raise ConfigurationFileInvalidError(
-                    f"Value for {key!r} ({value!r}) looks like an environment variable but has an invalid name."
+                    f"Missing value for expected environment variable {var_name!r} in config setting {config_key!r}"
                 )
-        if not value:
-            raise ConfigurationFileInvalidError(
-                f"Missing value for expected environment variable {var_name!r}"
-            )
-        interpolated[key] = value
-    return interpolated
+            config_value = config_value.replace(match_str, env_value)
```

Comment:
🤔 you are doing mutation in place here of the input `config_value` but you are also returning the `config_value`.
Could we avoid mutation and make this a more pure function instead? just returning new values?

Reply:
do you mean use a new variable to hold the interpolated value within the function? like...
```
new_value = None
if x:
    new_value = config_value.replace(foo,baa)
return new_value or config_value
```

Reply:
well, feel free to disagree with me here. I am not against mutations, but if we perform mutation, then the method should return nothing. It makes clear you are mutating the input. If you return a value, that value should be a new one, not a mutation of the input.
Where I find problems is when we have partial mutation behaviour: sometimes we mutate input, sometimes we return a new value.

That often leads to confusion and often is a source of accidents.

Reply:
wait that wouldn't work, we need to mutate the original `config_value` because it needs to retain multiple changes from different keys on successive iterations

Reply:
I may be misunderstanding the code. But is unclear why we need to do both, mutation and return of value.

Reply:
I think we do need to do both. The function `_interpolate_environment_vars` is performing multiple interpolations on a single string and then returning that string. If I were passing in the whole `yaml_data` dict, it would keep the mutated values, but with a single string argument, the function has to return the mutated string or the calling scope won't get the updated value. So it has to return `config_value`. And it's necessary to modify `config_value` in-place in the function, because if I assign `config_value.replace(x,y)` to a separate variable across multiple iterations, only the most recent key will be replaced in the final value.

Reply:
But, help me to understand something
In `main` you call this:

```
configuration = Configuration.from_file(args.configuration_path)
```

So, in the `from_file` method, you do 1 operation to do full replacement of all `${ENV_VAR}` in the parsed dict,
```
        try:
            yaml_data = path_to_yaml_data(path)
            yaml_data["site_settings"] = _interpolate_environment_vars(
                yaml_data["site_settings"]
            )
```
In there, rather than doing it only for `site_settings`, you can do the string replacement for everything that has an environment variable (we will need other properties, essentially, anything that's on the values side of the document may be useful to have support for env-vars).

From that point onwards, you can still run your validation steps. But you don't need to worry about environment variables anymore and you don't need to worry about having to mutate the configuration dict anymore.

Reply:
Right, the first version of `_interpolate_environment_vars` took a dict as input, because it was very tightly restricted by logic that related the name of the environment variable to the name of the config setting. So it was just iterating over the config settings available in `yaml_data['site_settings']`. Then some changes were necessary to allow it to a) use those same vars in other settings, and b) use multiple vars within a single setting. So in the latest version I changed it to accept a single config setting as a string (rather than a dict with multiple settings) and instead of the previous logic, predefined the names of acceptable environment variables and iterated over those.

With the new aim of matching any `${ENV_VAR}` formatted substring in any config setting, it will need to change again, and become a single pass over the entire `yaml_data` dict that is much less picky about acceptable values for env var names/config settings. So yes I guess it will move to a single operation that does all the replacements.